src/libzbxpgsql.c

```diff
@@ -567,3 +620,286 @@ char *strcat2(char *dest, const char *src)
+ * even with large numbers of keys.
+ *
+ * Returns: 
+ *   EXIT_SUCCESS = successfully inserted
+ *   EXIT_FAILURE = insert failed
+ *   -1           = duplicate key discarded
+ */
+int  storeSQLstmt(const char *key, const char *stmt) {
+    const char  *__function_name = "storeSQLstmt";
+    int  i;
+
+    zabbix_log(LOG_LEVEL_DEBUG, "%s: In %s(%s,<stmt>)", PACKAGE, __function_name, key);
+    // make sure we have space
+    if (SQLcount >= MAX_NUMBER_SQL_STATEMENT_IN_RAM) {
+        zabbix_log(LOG_LEVEL_ERR, "%s: ERROR: Keystore full: %i statements stored already", PACKAGE, SQLcount);
+        return EXIT_FAILURE;
+    }
+    // exclude dupes
+    if (SQLkeysearch((char *)key) != -1) {
+        zabbix_log(LOG_LEVEL_ERR, "%s: ERROR: Duplicate key: \"%s\"", PACKAGE, key);
+        return EXIT_FAILURE;
+    }
+    // start at the end of the index and push out
+    // entries to the next spot until you find the
+    // right spot to insert the new key/value pair
+    i = SQLcount - 1;
+    zabbix_log(LOG_LEVEL_TRACE, "%s: Starting to look for insert location (i:%i)", PACKAGE, i);
+    while (i >= 0 && strcmp(key,SQLkey[i]) < 0) {
+        zabbix_log(LOG_LEVEL_TRACE, "%s: moving data from slot %i to slot %i", PACKAGE, i, i+1);
+        SQLkey[i+1]  = SQLkey[i];
+        SQLstmt[i+1] = SQLstmt[i];
+        i--;
+    }
+    // allocate memory for the new key
+    SQLkey[i+1] = zbx_malloc(SQLkey[i+1],sizeof(char) * (strlen(key)+1));
+    if (SQLkey[SQLcount] == NULL) {
+        zabbix_log(LOG_LEVEL_CRIT, "%s: ERROR: zbx_malloc failed", PACKAGE);
+        return EXIT_FAILURE;
+    }
+    // allocate memory for the new value
+    SQLstmt[i+1] = zbx_malloc(SQLstmt[i+1],sizeof(char) * (strlen(stmt)+1));
+    if (SQLstmt[SQLcount] == NULL) {
+        zabbix_log(LOG_LEVEL_CRIT, "%s: ERROR: zbx_malloc failed", PACKAGE);
+        return EXIT_FAILURE;
+    }
+    // store the key and value
+    zabbix_log(LOG_LEVEL_TRACE, "%s: storing data in slot %i", PACKAGE, i+1);
+    zbx_strlcpy(SQLkey[i+1],key,strlen(key)+1);
+    zbx_strlcpy(SQLstmt[i+1],stmt,strlen(stmt)+1);
+    SQLcount++;
+
+    // Append NULL marker
+    SQLkey[SQLcount+1] = NULL;
+    SQLstmt[SQLcount+1] = NULL;
+
+    zabbix_log(LOG_LEVEL_DEBUG, "%s: End of %s", PACKAGE, __function_name);
+    return EXIT_SUCCESS;
+}
+
+/*
+ * Function SQLkeysearch
+ *
+ * Searches the key array to find the
+ * corresponding SQL stmt using binary
+ * search.
+ *
+ * Returns: 
+ *    If Key Found: index to key
+ *    If Not Found: -1
+ */
+int SQLkeysearch(char *key) {
+    const char  *__function_name = "SQLkeysearch";
+    int  top;
+    int  mid;
+    int  bottom;
+
+    zabbix_log(LOG_LEVEL_DEBUG, "%s: In %s(%s)", PACKAGE, __function_name, key);
+    top = SQLcount - 1;
+    bottom = 0;
+    while (bottom <= top) {
+        mid = (bottom + top)/2;
+        zabbix_log(LOG_LEVEL_TRACE, "%s: range top:%i mid:%i bottom:%i", PACKAGE, top, mid, bottom);
+        if (strcmp(SQLkey[mid], key) == 0) {
```

Comment:
I'll redo it.

On Sat, Aug 27, 2016 at 10:49 PM, Ryan Armstrong <notifications@github.com>
wrote:

> In src/libzbxpgsql.c
> <https://github.com/cavaliercoder/libzbxpgsql/pull/55#discussion_r76525684>
> :
>
> > + *    If Key Found: index to key
> > + *    If Not Found: -1
> > + */
> > +int SQLkeysearch(char *key) {
> > +    const char  *__function_name = "SQLkeysearch";
> > +    int  top;
> > +    int  mid;
> > +    int  bottom;
> > +
> > +    zabbix_log(LOG_LEVEL_DEBUG, "%s: In %s(%s)", PACKAGE, __function_name, key);
> > +    top = SQLcount - 1;
> > +    bottom = 0;
> > +    while (bottom <= top) {
> > +        mid = (bottom + top)/2;
> > +        zabbix_log(LOG_LEVEL_TRACE, "%s: range top:%i mid:%i bottom:%i", PACKAGE, top, mid, bottom);
> > +        if (strcmp(SQLkey[mid], key) == 0) {
>
> There's an opportunity here to call strcmp once and cache the result,
> instead of calling it three times on each iteration.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cavaliercoder/libzbxpgsql/pull/55/files/79848f2d016dd66c192d4d15c8fc91aae7a1d909#r76525684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOBRoPFqkWsaOHM52Nec-TF0iYU6Br2_ks5qkQVXgaJpZM4Ju1Cu>
> .
>